<a href="https://colab.research.google.com/github/MadgeneM/DDS8555/blob/main/MoiseMDDS8555_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Title: Assignment 2
# Author: Madgene Moise
# Date: Sunday, May 25, 2025

In [ ]:
# Install the ISLP library
!pip install ISLP

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 88.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.3/349.3 kB 26.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 522.0/522.0 kB 31.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.4/36.4 MB 50.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 83.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.1/823.1 kB 46.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 90.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 84.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 51.7 MB/s eta 0:00:00
  

In [ ]:
# Load the Carseats data set in Python
# Load necessary libraries and the Carseats dataset from seaborn-compatible source

import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

from ISLP import load_data
from ISLP.models import (ModelSpec as MS, summarize, poly)
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Fit a multiple regression model to predict Sales using Price, Urban, and US.


In [ ]:
# Load the Carseats dataset from ISLR via seaborn
carseats_data = load_data('Carseats')
#df = carseats_data
carseats_data.columns


Index(['Sales', 'CompPrice', 'Income', 'Advertising', 'Population', 'Price',
       'ShelveLoc', 'Age', 'Education', 'Urban', 'US'],
      dtype='object')

In [ ]:
# Variables: Sales (continuous), Price (continuous), Urban (Yes/No), US (Yes/No)

# Convert categorical variables
carseats_data['Urban'] = carseats_data['Urban'].astype('category')
carseats_data['US'] = carseats_data['US'].astype('category')


In [ ]:
# Fit the multiple regression model (a)
model_a = smf.ols('Sales ~ Price + Urban + US', data=carseats_data).fit()
model_a_summary = model_a.summary()

In [ ]:
# Return the summary of the model
model_a_summary.tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

**Interpretation:** When Price = 0, Urban = No, US = No, the predicted sales are 13.04 units (note: Price = 0 is unrealistic, so the intercept is mainly a baseline). Being in an urban areas increases sales by about 0.022 units, which is not statistically significant. Being located in the US increases sales by about 1.20 units, and this is statistically significant (p < 0.01). For each $1 increase in price, sales descrease by about 0.055 units, and this effect is highly significant (p < 0.001).

In [ ]:
# Fit a smaller model with only significant predictors
# Fit the multiple regression model (e)
model_e = smf.ols('Sales ~ Price + US', data=carseats_data).fit()
model_e_summary = model_e.summary()

Drop Urban since it was not siginifcant in the previous regression model.

In [ ]:
# Return the summary of the model (e)
model_e_summary.tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [ ]:
# Identify potential outliers or high leverage observations in model (e).
diagnostics_cleaned = model_e.get_influence().summary_frame()

In [ ]:
# Identify potential outliers (|standardized residual| > 2)
outliers = diagnostics_cleaned[diagnostics_cleaned["standard_resid"].abs() > 2]
outliers

dfb_Intercept  dfb_US[T.Yes]  dfb_Price   cooks_d  standard_resid  \
17       -0.060282       0.074948   0.063822  0.007122        2.109258   
25        0.239205      -0.165700  -0.176952  0.026109        2.581001   
28       -0.096722      -0.096306   0.102402  0.011054       -2.418794   
30        0.170894      -0.141670  -0.116566  0.015799        2.184464   
49       -0.099969      -0.168150   0.174387  0.022835        2.321404   
50       -0.050086      -0.108307   0.053028  0.011173       -2.811022   
57       -0.050586       0.159478  -0.015223  0.012972       -2.335540   
62        0.068379      -0.073995  -0.072395  0.007442       -2.096871   
68       -0.091104       0.092526   0.096454  0.011988        2.622678   
82       -0.090093       0.069639   0.095384  0.008290        2.015521   
98       -0.043119       0.075690   0.045651  0.006420        2.101526   
106       0.049093       0.153162  -0.115500  0.015429       -2.156279   
143       0.174987      -0.066058  -0.185264  0.016857       -2.052582   
209      -0.138953      -0.104454   0.147114  0.015663       -2.563363   
258      -0.198081       0.134017   0.147883  0.017677       -2.097292   
297      -0.057576      -0.086729   0.060958  0.007683       -2.230292   
298      -0.009376      -0.142836   0.070112  0.011532        2.045864   
304      -0.070264       0.071362   0.074391  0.007182        2.029897   
316       0.196637       0.093278  -0.208185  0.020470        2.169182   
328      -0.023883      -0.077556   0.025286  0.005623       -2.042436   
352      -0.024866       0.087167   0.026326  0.007552        2.376116   
376       0.143992       0.116257  -0.152448  0.018282        2.865082   
395      -0.048599       0.077420   0.051453  0.006908        2.156206   

     hat_diag  dffits_internal  student_resid    dffits  
17   0.004779         0.146167       2.118504  0.146807  
25   0.011622         0.279872       2.599652  0.281894  
28   0.005636        -0.182107      -2.433745 -0.183233  
30   0.009835         0.217708       2.194942  0.218753  
49   0.012553         0.261737       2.334376  0.263200  
50   0.004224        -0.183085      -2.835843 -0.184702  
57   0.007084        -0.197274      -2.348788 -0.198393  
62   0.005052        -0.149415      -2.105923 -0.150060  
68   0.005202         0.189645       2.642364  0.191069  
82   0.006085         0.157701       2.023359  0.158314  
98   0.004342         0.138775       2.110650  0.139378  
106  0.009857        -0.215143      -2.166285 -0.216141  
143  0.011861        -0.224878      -2.060961 -0.225796  
209  0.007100        -0.216768      -2.581586 -0.218309  
258  0.011913        -0.230286      -2.106350 -0.231280  
297  0.004612        -0.151815      -2.241569 -0.152582  
298  0.008198         0.185999       2.054143  0.186752  
304  0.005202         0.146781       2.037942  0.147363  
316  0.012883         0.247813       2.179402  0.248981  
328  0.004027        -0.129878      -2.050664 -0.130402  
352  0.003997         0.150520       2.390178  0.151410  
376  0.006637         0.234193       2.891521  0.236355  
395  0.004438         0.143957       2.166210  0.144625

Observations with standardized residuals beyond plus or minus 2 are considered potential outliers. Observations with standardized residuals less than -2 are index 258, 143, 209, 106, and 57. These values indicate unusually low Sales relative to what the model predicted. Furthermore, the Studentized residuals (adjusted for leverage) for these are also beyond plus or minus 2, confirming the concern.

In [ ]:
# Identify high leverage points (leverage > 2(p+1)/n), p=2 predictors, n ≈ 400
n = len(diagnostics_cleaned)
p = 2
leverage_threshold = 2 * (p + 1) / n
high_leverage = diagnostics_cleaned[diagnostics_cleaned["hat_diag"].astype(float) > leverage_threshold]
high_leverage

dfb_Intercept  dfb_US[T.Yes]  dfb_Price       cooks_d  standard_resid  \
42       -0.113501       0.030773   0.104206  4.329756e-03       -0.535475   
125      -0.066810       0.025165   0.058355  1.560768e-03       -0.419096   
155      -0.075466       0.039681   0.061155  2.218521e-03       -0.637633   
156      -0.072592      -0.089947   0.113027  7.855441e-03        1.229285   
159       0.004871      -0.002611  -0.003926  9.336799e-06        0.041896   
165       0.214047      -0.040295  -0.226618  1.975504e-02       -1.419635   
171       0.064174       0.023151  -0.067943  1.890242e-03        0.513988   
174       0.141260       0.093690  -0.185531  1.502431e-02       -1.213760   
191      -0.086372       0.023164   0.091444  3.553907e-03        0.761822   
203       0.058469       0.072448  -0.091038  5.103071e-03       -0.990793   
208      -0.092717       0.044481   0.076940  3.218758e-03       -0.721040   
269      -0.032289      -0.031065   0.046491  1.140355e-03        0.418099   
272       0.180889      -0.085270  -0.150750  1.212821e-02        1.382257   
313      -0.046356       0.018838   0.039907  7.644786e-04       -0.310986   
315      -0.065833       0.018456   0.069700  2.099193e-03        0.602567   
356       0.015440       0.015622  -0.022556  2.765330e-04       -0.211083   
365      -0.065676      -0.070092   0.097484  5.322958e-03        0.949655   
367       0.262743      -0.105127  -0.226888  2.428736e-02        1.732216   
383       0.001164      -0.000601  -0.000948  5.236050e-07        0.009672   
386      -0.026706      -0.030167   0.040345  9.462170e-04        0.410650   

     hat_diag  dffits_internal  student_resid    dffits  
42   0.043338        -0.113970      -0.534993 -0.113868  
125  0.025966        -0.068427      -0.418660 -0.068356  
155  0.016106        -0.081582      -0.637155 -0.081521  
156  0.015356         0.153513       1.230079  0.153612  
159  0.015707         0.005292       0.041843  0.005286  
165  0.028567        -0.243444      -1.421459 -0.243757  
171  0.021014         0.075304       0.513511  0.075234  
174  0.029687        -0.212304      -1.214486 -0.212431  
191  0.018039         0.103256       0.761418  0.103201  
203  0.015356        -0.123730      -0.990770 -0.123728  
208  0.018235        -0.098266      -0.720603 -0.098207  
269  0.019195         0.058490       0.417664  0.058429  
272  0.018687         0.190748       1.383849  0.190967  
313  0.023165        -0.047890      -0.310632 -0.047835  
315  0.017049         0.079357       0.602083  0.079294  
356  0.018279        -0.028803      -0.210829 -0.028768  
365  0.017399         0.126368       0.949537  0.126352  
367  0.023707         0.269930       1.736609  0.270614  
383  0.016514         0.001253       0.009660  0.001252  
386  0.016555         0.053279       0.410219  0.053223

leverage_threshold = 2 * (p + 1) / n
                   = 6/400 = 0.015.
All identified cases had leverage below the threshold (e.g., max at about 0.0286), so no observations are high leverage.